src/ffi.rs

```diff
@@ -524,39 +524,153 @@ pub extern "C" fn nrf_modem_os_is_in_isr() -> bool {
     cortex_m::peripheral::SCB::vect_active() != cortex_m::peripheral::scb::VectActive::ThreadMode
 }
 
-// This function is called by the library to allocate and initialize a mutex.
-// Required action:
-// Allocate and initialize a mutex.
-// If the address of an already allocated mutex is provided as an input, the allocation part is skipped and the mutex is only reinitialized.
-// Note
-// Mutexes are not required if multithreaded access to modem functionalities is not needed. In this case, the function must blindly return 0.
+// A basic mutex lock implementation for the os mutex functions below
+struct MutexLock {
+    lock: AtomicBool
+}
+
+impl MutexLock {
+    pub fn lock(&self) -> bool {
+        match self.lock.compare_exchange(false, true, Ordering::SeqCst, Ordering::SeqCst) {
+            Ok(false) => true,
+            _ => false
+        }
+    }
+
+    pub fn unlock(&self) {
+        self.lock.store(false, Ordering::SeqCst);
+    }
+}
+
+/// Initialize a mutex.
+///
+/// The function shall allocate and initialize a mutex and return its address
+/// as an output. If an address of an already allocated mutex is provided as
+/// an input, the allocation part is skipped and the mutex is only reinitialized.
+///
+/// **Parameters**:
+/// - mutex – (inout) The address of the mutex.
+///
+/// **Returns**
+/// - 0 on success, a negative errno otherwise.
 #[no_mangle]
-pub unsafe extern "C" fn nrf_modem_os_mutex_init() -> i32 {
-   // TODO FIXME 
-   0
+pub unsafe extern "C" fn nrf_modem_os_mutex_init(
+    mutex: *mut *mut core::ffi::c_void
+) -> core::ffi::c_int {
+    if mutex.is_null() {
+        return -(nrfxlib_sys::NRF_EINVAL as i32);
+    }
+
+    // Allocate if we need to
+    if (*mutex).is_null() {
+        // Allocate our mutex datastructure
+        *mutex = nrf_modem_os_alloc(core::mem::size_of::<MutexLock>()) as *mut _;
+
+        if (*mutex).is_null() {
+            // We are out of memory
+            return -(nrfxlib_sys::NRF_ENOMEM as i32);
+        }
+    } 
+
+    // (Re-)Initialize (unlock) the mutex
+    (*(mutex as *mut MutexLock)).unlock();
```

Comment:
Technically speaking the `MutexLock` is uninitialized.
I think it'd be better to put it in a `MaybeUninit` and write a fresh properly initialized `MutexLock` into it.

Does it really need to? Well... maybe? I don't know if `AtomicBool` is fine with random data. I think it's UB for a normal bool.
But even if it is fine, using `MaybeUninit` is more proper.

Reply:
That is right.  Updated.